File: Programmers/92341.py
import math

def solution(fees, records):
    in_cars = {}    # 현재 주차장에 있는 차
    durations = {}  # 차별 누적 주차 시간
    prices = {}  # 차별 누적된 요금
    
    for record in records:
        time, car, move = map(str, record.split())
        
        if move == "IN":
            in_cars[car] = time
            
        else:
            in_h, in_m = map(int, in_cars[car].split(':'))
            out_h, out_m = map(int, time.split(':'))
            
            del in_cars[car]
            
            # 주차 시간 구하기
            if out_m >= in_m:
                duration = (out_h - in_h) * 60 + (out_m - in_m)
            else:
                duration = (out_h - in_h - 1) * 60 + (out_m + 60 - in_m)
                
            if car in durations:
                durations[car] += duration
            else:
                durations[car] = duration
                
    for car in in_cars:
        in_h, in_m = map(int, in_cars[car].split(':'))
        duration = (23 - in_h) * 60 + (59 - in_m)
        if car in durations:
            durations[car] += duration
        else:
            durations[car] = duration
    
    # 요금 구하기
    for car in durations:
        if durations[car] > fees[0]:
            price = fees[1] + math.ceil((durations[car] - fees[0]) / fees[2]) * fees[3]
        else:
            price = fees[1]
        prices[car] = price
                
    prices = sorted(prices.items())
    answer = [price[1] for price in prices]
    return answer

File: Programmers/92341.js
function solution(fees, records) {
  let inouts = {};
  let timeSum = {};

  records.forEach((record) => {
    let [time, car, type] = record.split(" ");
    let [h, m] = time.split(":").map(Number);

    if (type === "IN") inouts[car] = h * 60 + m;
    else {
      if (timeSum[car]) timeSum[car] += h * 60 + m - inouts[car];
      else timeSum[car] = h * 60 + m - inouts[car];
      delete inouts[car];
    }
  });

  for (let car in inouts) {
    const out = 23 * 60 + 59;
    if (timeSum[car]) timeSum[car] += out - inouts[car];
    else timeSum[car] = out - inouts[car];
  }

  let prices = [];
  for (let car in timeSum) {
    if (timeSum[car] <= fees[0]) prices.push([car, fees[1]]);
    else
      prices.push([
        car,
        fees[1] + Math.ceil((timeSum[car] - fees[0]) / fees[2]) * fees[3],
      ]);
  }

  prices.sort((a, b) => a[0] - b[0]);
  return prices.map((p) => p[1]);
}
